scripts/smoke-test-droid.ps1

```diff
@@ -58,37 +58,54 @@ foreach ($device in $DeviceList)
     {
         Throw "Failed to Install APK: $stdout."
     }
-    $AppStarted = 'False'
 
-    Write-Output "Clearing logcat from $device."
-    adb -s $device logcat -c
+    $FlakyRetry = 3
+    While ($FlakyRetry -gt 0)
+    {
+        $AppStarted = 'False'
 
-    Write-Output "Starting Test..."
+        Write-Output "Clearing logcat from $device."
+        adb -s $device logcat -c
 
-    adb -s $device shell am start -n $TestActivityName -e test smoke
-    #despite calling start, the app might not be started yet.
+        Write-Output "Starting Test..."
 
-    $Timeout = 30
-    While ($Timeout -gt 0) 
-    {
-        #Get a list of active processes
-        $processIsRunning = (adb -s $device shell ps)
-        #And filter by ProcessName
-        $processIsRunning = $processIsRunning | select-string $ProcessName
-        
-        If ($processIsRunning -eq $null -And $AppStarted -eq 'True')
+        adb -s $device shell am start -n $TestActivityName -e test smoke
+        #despite calling start, the app might not be started yet.
+
+        $Timeout = 30
+        While ($Timeout -gt 0) 
+        {
+            #Get a list of active processes
+            $processIsRunning = (adb -s $device shell ps)
+            #And filter by ProcessName
+            $processIsRunning = $processIsRunning | select-string $ProcessName
+            
+            If ($processIsRunning -eq $null -And $AppStarted -eq 'True')
+            {
+                $Timeout = -2
+                break
+            }
+            ElseIf ($processIsRunning -ne $null -And $AppStarted -eq 'False')
+            {
+                # Some devices might take a while to start the test, so we wait for the activity to start before checking if it was closed.
+                $AppStarted = 'True'
+            }
+            Write-Output "Waiting Process on $device to complete, waiting $Timeout seconds"
+            Start-Sleep -Seconds 1
+            $Timeout--
+        }
+
+        $stdout = adb -s $device logcat -d  | select-string 'Unity   : Timeout while trying detaching'
+        If ($stdout -eq $null)
         {
-            $Timeout = -2
             break
         }
-        ElseIf ($processIsRunning -ne $null -And $AppStarted -eq 'False')
+        Else
         {
-            # Some devices might take a while to start the test, so we wait for the activity to start before checking if it was closed.
-            $AppStarted = 'True'
+            Write-Warning "Test was flaky, retrying."
+            $FlakyRetry--
+            Start-Sleep -Seconds 3
```

Comment:
Can you really just restart it like that? Is it dead and stuck and hitting the timeout because of that? You can run the process twice? Am I missing the cleanup?

Reply:
The app is already closed when entering this if condition. So all it does is open the test app again and repeat the tests.
What's happening seems to be Unity is killing the app before even running the tests.